### DISCUSSION TITLE
Generate random string in template

### DISCUSSION THREAD

**Discussion by miquecg** [twpayne/chezmoi#2624] [Q&A] [answered]
Would it be possible to generate a random string using template functions? So far I didn't find any helper, but I've thought that maybe calling a shell function would do the trick.

> $ chezmoi execute-template '{{ run "echo foo" }}'

But `run` and `exec` seem to not be available.

**Comment by halostatue** [comment]
The function you’re looking for is `output`: https://www.chezmoi.io/reference/templates/functions/output/

**Reply by miquecg** [reply]
Thanks for the pointer. It looks like it has some limitations though and commands cannot be piped together, or am I missing something? As a workaround I could wrap the whole thing with a script.

For the record I'm trying to achieve something like:

> LC_ALL=C tr -dc A-Za-z </dev/urandom | head -c 13

**Reply by bradenhilton** [reply]
If you want to pipe within the command you need to use `output "sh" "-c" "... | ..."`.

`{{ output "sh" "-c" "LC_ALL=C tr -dc A-Za-z </dev/urandom | head -c 13" }}`

**Reply by miquecg** [reply]
That's what I assumed, either everything inline or call a script.

Thanks.

**Comment by 2-4601** [comment, marked as answer]
https://masterminds.github.io/sprig/strings.html#randalphanum-randalpha-randnumeric-and-randascii

> These four functions generate cryptographically secure (uses `crypto/rand`) random strings, but with different base character sets:
> 
> `randAlphaNum` uses `0-9a-zA-Z`
> `randAlpha` uses `a-zA-Z`
> `randNumeric` uses `0-9`
> `randAscii` uses all printable ASCII characters
> Each of them takes one parameter: the integer length of the string.
> 
> `randNumeric 3`
> 
> The above will produce a random string with three digits.
```
$ chezmoi execute-template '{{ randAlpha 10 }}'
yAcuBuYIpo
```
----------

https://masterminds.github.io/sprig/crypto.html#randbytes
> The `randBytes` function accepts a count `N` and generates a cryptographically secure (uses `crypto/rand`) random sequence of `N` bytes. The sequence is returned as a base64 encoded string.
```
$ chezmoi execute-template '{{ randBytes 16 }}'
txZKVdlnWbjddmApbBTVvA==
```
----------

https://masterminds.github.io/sprig/uuid.html#uuid-functions
> Sprig can generate UUID v4 universally unique IDs.
> 
> `uuidv4`
>
> The above returns a new UUID of the v4 (randomly generated) type.

```
$ chezmoi execute-template '{{ uuidv4 }}'
a97772b8-d001-4220-847f-36d3a80057d3
```

**Reply by halostatue** [reply]
Depending on the type of random value needed, this is a better overall answer.

**Reply by miquecg** [reply]
Oh, great! Didn't know about those helper functions :)
